### PR TITLE
[RTR] QandQdotBounds Issues

### PR DESCRIPTION
**Pull request by Kev1CO**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document [docs/contribution.md]?
* [x] Have you checked to ensure there aren't other open [Pull Requests] for the same update/change?
* [x] Have you opened/linked the issue related to your pull request?
* [x] Have you used the tag [WIP] for on-going changes, and removed it when the pull request was ready?
* [ ] When ready to merge, have you sent a comment pinging @pariterre in it?

### New Feature Submissions:

1. [x] Does your submission pass the tests (if not please explain why this is intended)?
2. [x] Did you write a proper documentation (docstrings and ReadMe)
3. [x] Have you linted your code locally prior to submission (using the command: `black . -l120 --exclude "external/*"`)?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new examples for your core changes, as applicable?
* [x] Have you written new tests for your core changes, as applicable?
Related to #587

- - -
This change is [Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)

### CODE REVIEW

File: bioptim/interfaces/biomodel.py [hunk 1]
```diff
@@ -192,3 +192,39 @@ def contact_forces(self, q, qdot, tau, external_forces: list = None) -> MX:
...
+    def q_range(self) -> list:
+        """Get the q_ranges"""
+
+    def qdot_range(self) -> list:
+        """Get the qdot_ranges"""
+
+    def qddot_range(self) -> list:
+        """Get the qddot_ranges"""
+
+    def q_and_q_dot_range(self, index_to_keep):
```
Comment (pariterre) [inline bioptim/interfaces/biomodel.py:205]: This method does not make much sense in an API point of view. We already provide q and qdot ranges, why would we need to supply q_and_qdot too?

File: bioptim/interfaces/biomodel.py [hunk 2]
```diff
@@ -192,3 +192,39 @@ def contact_forces(self, q, qdot, tau, external_forces: list = None) -> MX:
...
+        Parameters
+        ----------
+        index_to_keep: BiMappingList
+            The mapping of q and qdot (if only q, then qdot = q)
+
+        Returns
+        -------
+        The minimum and maximum of the x_bound
+        """
+    def q_and_q_dot_and_q_ddot_range(self, index_to_keep):
```
Comment (pariterre) [inline bioptim/interfaces/biomodel.py:218]: Same thing. Also, please do not separate the dot with its relative dotted, it is unreadable

File: bioptim/interfaces/biorbd_model.py [hunk 3]
```diff
@@ -299,42 +299,42 @@ def contact_forces(self, q, qdot, tau, external_forces: list = None) -> MX:
...
+
+    def q_and_q_dot_range(self, index_to_keep):
+        q_ranges = self.q_range()
+        qdot_ranges = self.qdot_range()
+        x_min = [q_ranges[i].min() for i in index_to_keep["q"].to_first.map_idx] + [
+            qdot_ranges[i].min() for i in index_to_keep["qdot"].to_first.map_idx]
+        x_max = [q_ranges[i].max() for i in index_to_keep["q"].to_first.map_idx] + [
+            qdot_ranges[i].max() for i in index_to_keep["qdot"].to_first.map_idx
+        ]
+        print(index_to_keep)
```
Comment (pariterre) [inline bioptim/interfaces/biorbd_model.py:331]: To be removed

**Review by pariterre — CHANGES_REQUESTED** [review]

File: bioptim/interfaces/biomodel.py [hunk 4]
```diff
@@ -192,3 +192,39 @@ def contact_forces(self, q, qdot, tau, external_forces: list = None) -> MX:
 
     def passive_joint_torque(self, q, qdot) -> MX:
         """Get the passive joint torque"""
+
+    def q_range(self) -> list:
+        """Get the q_ranges"""
```
[the remark this answers is not in the document]
Comment (pariterre) [reply bioptim/interfaces/biomodel.py:197]: float

**Comment by pariterre** [comment]
Please comment "Done" in each message when it is Done :)

**Review by Kev1CO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)* status: 0 of 5 files reviewed, 6 unresolved discussions (waiting on @pariterre)

___
*[`bioptim/interfaces/biomodel.py` line 197 at r4](https://reviewable.io/reviews/pyomeca/bioptim/604#-NM_rqAsCIAlw-rjmn0E-r4-197:-NNDzhec0arTNfbzVYX5:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/c2910d057a0f4d9aefac1139a23e34fccc24e1c6/bioptim/interfaces/biomodel.py#L197)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

float
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biomodel.py` line 205 at r4](https://reviewable.io/reviews/pyomeca/bioptim/604#-NM_rqAsCIAlw-rjmn0E-r4-205:-NNDzNIS01JzvnImDEO6:b-sx84ma) ([raw file](https://github.com/pyomeca/bioptim/blob/c2910d057a0f4d9aefac1139a23e34fccc24e1c6/bioptim/interfaces/biomodel.py#L205)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

This method does not make much sense in an API point of view. We already provide q and qdot ranges, why would we need to supply q_and_qdot too?
</blockquote></details>

obsolete
___
*[`bioptim/interfaces/biomodel.py` line 218 at r4](https://reviewable.io/reviews/pyomeca/bioptim/604#-NM_rqAsCIAlw-rjmn0E-r4-218:-NNDzk5M9m7amvg4iRHI:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/c2910d057a0f4d9aefac1139a23e34fccc24e1c6/bioptim/interfaces/biomodel.py#L218)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Same thing. Also, please do not separate the dot with its relative dotted, it is unreadable
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biorbd_model.py` line 331 at r4](https://reviewable.io/reviews/pyomeca/bioptim/604#-NM_Ry6TsAb3GFmAxedS-r4-331:-NNE0eLMDs6G33iIiBNb:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/c2910d057a0f4d9aefac1139a23e34fccc24e1c6/bioptim/interfaces/biorbd_model.py#L331)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

To be removed
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biorbd_model.py` line 338 at r4](https://reviewable.io/reviews/pyomeca/bioptim/604#-NM_Ry6TsAb3GFmAxedS-r4-338:-NNE0mdF3u3iQoiH3Lg9:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/c2910d057a0f4d9aefac1139a23e34fccc24e1c6/bioptim/interfaces/biorbd_model.py#L338)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

As stated prior, please remove these two methods and let the Q and Q dot bound concatenate what it needs by itself
</blockquote></details>

Done.
___
*[`bioptim/limits/path_conditions.py` line 644 at r4](https://reviewable.io/reviews/pyomeca/bioptim/604#-NM_Ry6TsAb3GFmAxedT-r4-644:-NNE0zn60zorj06ootUm:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/c2910d057a0f4d9aefac1139a23e34fccc24e1c6/bioptim/limits/path_conditions.py#L644)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

At this point, I feel a `states` variable would better fit our needs: for instance `StatesBounds(states=["q", "qdot", ...])`. In the biomod, instead of `q_bounds` and the others, it should have `state_bounds(state)` and it returns the relevant bound for the required state
</blockquote></details>

Done.

File: bioptim/interfaces/biomodel.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kev1CO) [reply bioptim/interfaces/biomodel.py]: Outdated

File: bioptim/interfaces/biomodel.py (hunk 1 above)
Comment (Kev1CO) [reply bioptim/interfaces/biomodel.py]: Outdated

File: bioptim/interfaces/biomodel.py (hunk 2 above)
Comment (Kev1CO) [reply bioptim/interfaces/biomodel.py]: Done

File: bioptim/interfaces/biorbd_model.py (hunk 3 above)
Comment (Kev1CO) [reply bioptim/interfaces/biorbd_model.py]: Done

File: bioptim/interfaces/biorbd_model.py [hunk 5]
```diff
@@ -299,42 +299,42 @@ def contact_forces(self, q, qdot, tau, external_forces: list = None) -> MX:
...
+        return x_min, x_max
+
+    def q_and_q_dot_and_q_ddot_range(self, index_to_keep):
+        qddot_ranges = self.qddot_range()
+        x_min = [qddot_ranges[i].min() for i in index_to_keep["qddot"].to_first.map_idx]
+        x_max = [qddot_ranges[i].max() for i in index_to_keep["qddot"].to_first.map_idx]
+        return x_min, x_max
```
[the remark this answers is not in the document]
Comment (Kev1CO) [reply bioptim/interfaces/biorbd_model.py]: Done

File: bioptim/limits/path_conditions.py [hunk 6]
```diff
@@ -638,26 +638,8 @@ def __init__(
...
-        #
-        # super(QAndQDotBounds, self).__init__(min_bound=x_min, max_bound=x_max)
-
-#         MODIF
-#         print(bio_model.dof_mappings)
-        super(QAndQDotBounds, self).__init__(min_bound=bio_model.q_and_qdot_bounds_x_min(dof_mappings), max_bound=bio_model.q_and_qdot_bounds_x_max(dof_mappings))
+        x_min, x_max = bio_model.q_and_q_dot_range(dof_mappings)
+        super(QAndQDotBounds, self).__init__(min_bound=x_min, max_bound=x_max)
 
 class QAndQDotAndQDDotBounds(QAndQDotBounds):
```
[the remark this answers is not in the document]
Comment (Kev1CO) [reply bioptim/limits/path_conditions.py]: Done

**Review by pariterre — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r5, 2 of 4 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)* status: all files reviewed, 7 unresolved discussions (waiting on @Kev1CO)

___
*[`bioptim/interfaces/biomodel.py` line 197 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIZ-sV1tF--BZzQGGq:-NNIZ-sV1tF--BZzQGGr:b-1lr62m) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biomodel.py#L197)):*
> ```Python
>         """Get the passive joint torque"""
> 
>     def bounds_from_ranges(self, variables: str | list[str, ...], mapping: BiMapping | BiMappingList = None) -> "Bounds":
> ```

After testing, it seems importing `Bounds` is now working
___
*[`bioptim/interfaces/biorbd_model.py` line 9 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIXKvOA1LScW8NhhyL:-NNIXKvOA1LScW8NhhyM:b-lqr870) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L9)):*
> ```Python
> import numpy as np
> 
> from bioptim.misc.utils import check_version
> ```

Please use relative imports
___
*[`bioptim/interfaces/biorbd_model.py` line 307 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIXwc89nJHqMLuKoSK:-NNIXwc89nJHqMLuKoSL:bertu2h) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L307)):*
> ```Python
>         return self.model.passiveJointTorque(q, qdot).to_mx()
> 
>     def q_mapping(self, mapping: BiMapping = None) -> BiMapping:
> ```

I feel that this should be a protected method.. (`_q_mapping`)
___
*[`bioptim/interfaces/biorbd_model.py` line 323 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIYh7wBipwdc5uGNSQ:-NNIYh7x0IJfo_FBo5DD:b-1nnkkg) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L323)):*
> ```Python
>         return mapping
> 
>     def qdot_mapping(self, mapping: BiMapping = None) -> BiMapping:
> ```

I feel that this should be a protected method.. (`_qdot_mapping`)
___
*[`bioptim/interfaces/biorbd_model.py` line 335 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIYiCg9jTW6ogjI4mg:-NNIYiCg9jTW6ogjI4mh:b-pgxtk2) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L335)):*
> ```Python
>         return mapping
> 
>     def qddot_mapping(self, mapping: BiMapping = None) -> BiMapping:
> ```

I feel that this should be a protected method.. (`_qddot_mapping`)
___
*[`bioptim/interfaces/biorbd_model.py` line 370 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIXc0Y5MkmuSNac3i7:-NNIXc0Y5MkmuSNac3i8:bnpe8r) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L370)):*
> ```Python
>                     out.concatenate(Bounds(min_bound=x_min, max_bound=x_max))
> 
>         for var in variables:
> ```

This `for` loop is not necessary (it can be done in the previous for)
___
*[`bioptim/interfaces/biorbd_model.py` line 382 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIXov_6SHFGVuePfbc:-NNIXov_6SHFGVuePfbd:bnpe8r) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L382)):*
> ```Python
>                     out.concatenate(Bounds(min_bound=x_min, max_bound=x_max))
> 
>         for var in variables:
> ```

This `for` loop is not necessary (it can be done in the previous for)

**Review by Kev1CO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)* status: 4 of 5 files reviewed, 7 unresolved discussions (waiting on @pariterre)

___
*[`bioptim/interfaces/biomodel.py` line 197 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIZ-sV1tF--BZzQGGq:-NNghklsD-3P0E1aws2Y:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biomodel.py#L197)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

After testing, it seems importing `Bounds` is now working
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biorbd_model.py` line 9 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIXKvOA1LScW8NhhyL:-NNghlVZDs6aOkIBCsMT:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L9)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Please use relative imports
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biorbd_model.py` line 307 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIXwc89nJHqMLuKoSK:-NNghluZ85MT9_gZZeCi:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L307)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

I feel that this should be a protected method.. (`_q_mapping`)
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biorbd_model.py` line 323 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIYh7wBipwdc5uGNSQ:-NNghmEA1B0Hm4QqlUOf:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L323)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

I feel that this should be a protected method.. (`_qdot_mapping`)
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biorbd_model.py` line 335 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIYiCg9jTW6ogjI4mg:-NNghmZf4uzUYhH3L-1N:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L335)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

I feel that this should be a protected method.. (`_qddot_mapping`)
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biorbd_model.py` line 370 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIXc0Y5MkmuSNac3i7:-NNghn5U6-D9oMRF5UwI:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L370)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

This `for` loop is not necessary (it can be done in the previous for)
</blockquote></details>

Done.
___
*[`bioptim/interfaces/biorbd_model.py` line 382 at r8](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNIXov_6SHFGVuePfbc:-NNghnJQ5IbNPLy2aF-G:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/bf5154c43c37a966eba5911268f74d39efbfdb1e/bioptim/interfaces/biorbd_model.py#L382)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

This `for` loop is not necessary (it can be done in the previous for)
</blockquote></details>

Done.

**Review by Kev1CO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)* status: 88 of 89 files reviewed, 2 unresolved discussions (waiting on @pariterre)

___
*[`bioptim/interfaces/biomodel.py` line 205 at r10](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNlJUgfCDMT_2HCW4M3:-NNqcQgo4FCzxVrvYqlP:b-w1y986) ([raw file](https://github.com/pyomeca/bioptim/blob/cd4560bef0b8a62f7552a2d1a6ecedf9eaed278a/bioptim/interfaces/biomodel.py#L205)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

We don't know in "BioModel" which states exists in the model. It could basically be anything for CustomModels
</blockquote></details>

Not sure to understand
___
*[`bioptim/interfaces/biorbd_model.py` line 354 at r10](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNlJkIxADQTLAXKGW4K:-NNqcU698Lxkkl-TR-kh:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/cd4560bef0b8a62f7552a2d1a6ecedf9eaed278a/bioptim/interfaces/biorbd_model.py#L354)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

Use relative imports
Also, if "Bounds" is an Output, it probably means it is already imported
</blockquote></details>

Done.

**Review by pariterre — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)* status: all files reviewed, 1 unresolved discussion (waiting on @Kev1CO)

___
*[`bioptim/interfaces/biomodel.py` line 205 at r10](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNlJUgfCDMT_2HCW4M3:-NNrTkOOAHS_GGtKdj-e:b4nq6ti) ([raw file](https://github.com/pyomeca/bioptim/blob/cd4560bef0b8a62f7552a2d1a6ecedf9eaed278a/bioptim/interfaces/biomodel.py#L205)):*
<details><summary><i>Previously, Kev1CO wrote…</i></summary><blockquote>

Not sure to understand
</blockquote></details>

such as

**Review by pariterre — COMMENTED** [review]
> Reviewed 26 of 26 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)* status: all files reviewed, 1 unresolved discussion (waiting on @Kev1CO)

**Review by Kev1CO — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)* status: 87 of 89 files reviewed, 1 unresolved discussion (waiting on @pariterre)

___
*[`bioptim/interfaces/biomodel.py` line 205 at r10](https://reviewable.io/reviews/pyomeca/bioptim/604#-NNlJUgfCDMT_2HCW4M3:-NNsGr9SCWQJvgmiaxkp:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/cd4560bef0b8a62f7552a2d1a6ecedf9eaed278a/bioptim/interfaces/biomodel.py#L205)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

such as
</blockquote></details>

Done.

**Review by pariterre — APPROVED** [review]
> Reviewed 2 of 2 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Kev1CO)

**Review by pariterre — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/604#-:-NNvdh_q0S05QRo-SMyA:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Kev1CO)

**Comment by pariterre** [comment]
GG

**Pull request merged**